src/controller/InteractionPeer.h

```diff
@@ -0,0 +1,100 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <app/CommandSender.h>
+#include <app/InteractionModelEngine.h>
+#include <app/util/CHIPDeviceCallbacksMgr.h>
+#include <app/util/basic-types.h>
+#include <controller/data_model/gen/CHIPClientCallbacks.h>
+
+namespace chip {
+namespace Controller {
+
+class InteractionPeerDelegate
+{
+public:
+    virtual CHIP_ERROR PreparePeer() = 0;
+
+    virtual NodeId GetDeviceId() const = 0;
+
+    virtual FabricIndex GetFabricIndex() const = 0;
+
+    virtual SecureSessionHandle * GetSesssionHandle() = 0;
+
+    virtual CHIP_ERROR SendMessage(Protocols::Id protocolId, uint8_t msgType, Messaging::SendFlags sendFlags,
+                                   System::PacketBufferHandle && message) = 0;
+
+    virtual ~InteractionPeerDelegate() {}
+};
+
+class InteractionPeer
+{
+public:
+    InteractionPeer(InteractionPeerDelegate * delegate) : mDelegate(delegate) {}
+
+    CHIP_ERROR SendMessage(Protocols::Id protocolId, uint8_t msgType, Messaging::SendFlags sendFlags,
+                           System::PacketBufferHandle && message);
+
+    template <typename MessageType, typename = std::enable_if_t<std::is_enum<MessageType>::value>>
+    CHIP_ERROR SendMessage(MessageType msgType, Messaging::SendFlags sendFlags, System::PacketBufferHandle && message)
+    {
+        return SendMessage(Protocols::MessageTypeTraits<MessageType>::ProtocolId(), to_underlying(msgType), sendFlags,
+                           std::move(message));
+    }
+
+    CHIP_ERROR SendReadAttributeRequest(app::AttributePathParams aPath, Callback::Cancelable * onSuccessCallback,
+                                        Callback::Cancelable * onFailureCallback, app::TLVDataFilter aTlvDataFilter);
+
+    CHIP_ERROR SendWriteAttributeRequest(app::WriteClientHandle aHandle, Callback::Cancelable * onSuccessCallback,
+                                         Callback::Cancelable * onFailureCallback);
+
+    uint8_t GetNextSequenceNumber() { return mSequenceNumber++; };
```

Comment:
Need to explain what this class is meant to do, and what the different methods do.

Reply:
Doc string added. These functions are moved from the `Device` class. I'm not very sure about the functionality of some methods. TODOs added for the origional author @erjiaqing to add doc strings.